tests/test_server.py

```diff
@@ -59,17 +59,125 @@ def udev_rules(self):
     def client(self):
         yield Client()
 
-    def add_device(self, client, name='steelseries-rival310'):
+    def add_device(self, client, spec={}):
         data = {
-            'shortname': name
+            'driver': 'generic',
+            'specification': {
+                'bus_info': [0x1, 0x0001, 0x0001],
+                'name': 'Generic Device',
+                'shortname': 'generic',
+                'report_descriptors': {
+                    0: [
+                        0x05, 0x01,  # .Usage Page (Generic Desktop)        0
+                        0x09, 0x02,  # .Usage (Mouse)                       2
+                        0xa1, 0x01,  # .Collection (Application)            4
+                        0x09, 0x02,  # ..Usage (Mouse)                      6
+                        0xa1, 0x02,  # ..Collection (Logical)               8
+                        0x09, 0x01,  # ...Usage (Pointer)                   10
+                        0xa1, 0x00,  # ...Collection (Physical)             12
+                        0x05, 0x09,  # ....Usage Page (Button)              14
+                        0x19, 0x01,  # ....Usage Minimum (1)                16
+                        0x29, 0x03,  # ....Usage Maximum (3)                18
+                        0x15, 0x00,  # ....Logical Minimum (0)              20
+                        0x25, 0x01,  # ....Logical Maximum (1)              22
+                        0x75, 0x01,  # ....Report Size (1)                  24
+                        0x95, 0x03,  # ....Report Count (3)                 26
+                        0x81, 0x02,  # ....Input (Data,Var,Abs)             28
+                        0x75, 0x05,  # ....Report Size (5)                  30
+                        0x95, 0x01,  # ....Report Count (1)                 32
+                        0x81, 0x03,  # ....Input (Cnst,Var,Abs)             34
+                        0x05, 0x01,  # ....Usage Page (Generic Desktop)     36
+                        0x09, 0x30,  # ....Usage (X)                        38
+                        0x09, 0x31,  # ....Usage (Y)                        40
+                        0x15, 0x81,  # ....Logical Minimum (-127)           42
+                        0x25, 0x7f,  # ....Logical Maximum (127)            44
+                        0x75, 0x08,  # ....Report Size (8)                  46
+                        0x95, 0x02,  # ....Report Count (2)                 48
+                        0x81, 0x06,  # ....Input (Data,Var,Rel)             50
+                        0xc0,        # ...End Collection                    52
+                        0xc0,        # ..End Collection                     53
+                        0xc0,        # .End Collection                      54
+                    ]
+                },
+            }
```

Comment:
This above should all be handled in the server, not in the client.
I do not want from libratbag or curl to have to type in the report descriptor, or the VID/PID.

All of these infromation are known before hand either in the `BaseDevice` class, either in each protocol.

Reply:
How? How does `HIDPP20Device` know the VID/PID? Do we have config files for each device?

Reply:
see my code example above - yes - yes (but we can hardcode anything in the code)